### PR TITLE
backends: try only utf8 when UTF8ONLY

### PR DESCRIPTION
**Pull request by half-duplex**
### Description
When UTF8ONLY is active, raise a decoding error if the message fails to decode instead of attempting other encodings.

As far as I can tell, Sopel never sends non-utf8 data unless someone's poking at the backend directly, so no other changes should be necessary to consider UTF8ONLY supported. (Actually, even this doesn't seem mandatory, just sensible)

### Checklist
- [x] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [x] I can and do license this contribution under the EFLv2
- [x] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [x] I have tested the functionality of the things this change touches

### CODE REVIEW

**Comment by half-duplex** [comment]
Was trying to add tests and... Windows-1252 is a superset of ISO/IEC 8859-1. I'm removing the latter.

File: sopel/irc/abstract_backends.py [hunk 1]
```diff
@@ -67,16 +67,15 @@ def decode_line(self, line: bytes) -> str:
         # We can't trust clients to pass valid Unicode.
         try:
             data = str(line, encoding='utf-8')
-        except UnicodeDecodeError:
+        except UnicodeDecodeError as e:
+            # ...unless the server announces UTF8ONLY
+            if "UTF8ONLY" in self.isupport:
+                raise e
```
Comment (dgw) [inline sopel/irc/abstract_backends.py:73]: Wow, this is the perfect place for `e.add_note("UTF8ONLY server sent a line that could not be decoded as UTF-8")`, but that requires Python 3.11 🙁 (or a nasty monkeypatch, I guess)

**Review by dgw — COMMENTED** [review]
No test case with invalid input that will trigger the final `raise ValueError`?

File: test/irc/test_irc_abstract_backends.py [hunk 2]
```diff
@@ -298,3 +300,31 @@ def test_send_notice_safe():
...
+    assert backend.decode_line(test.encode("utf-8")) == test
+
+
+def test_decode_line_utf8only():
+    bot = BotCollector()
+    backend = MockIRCBackend(bot)
+    backend.isupport = ISupport(UTF8ONLY=None)
+
+    test = "PRIVMSG #sopel :Hello, Martín!"
+    with raises(UnicodeDecodeError):
```
Comment (dgw) [inline test/irc/test_irc_abstract_backends.py:320]: I tend to prefer `with pytest.raises` because it's clearer. Others (Exi) might disagree, but I like to keep the namespace of imported stuff. See-also: My preference for e.g. `from sopel import plugin` and then `@plugin.command()`.

(Forgot to add this _before_ submitting that review just now, sorry.)

File: sopel/irc/abstract_backends.py (hunk 1 above)
Comment (half-duplex) [reply sopel/irc/abstract_backends.py:73]: Yeah. Not a bad place for a regular LOGGER.warning() though.

Comment (half-duplex) [reply sopel/irc/abstract_backends.py:73]: Never mind, there's no LOGGER in there... meh

File: test/irc/test_irc_abstract_backends.py (hunk 2 above)
Comment (half-duplex) [reply test/irc/test_irc_abstract_backends.py]: Yep, fair. It's also tests so being thrifty with includes isn't exactly priority zero.

**Comment by half-duplex** [comment]
Added, but more for coverage than because it matters =p

**Review by dgw — APPROVED** [review]
Cool. Assuming CI passes and all that (why wouldn't it?), it's squash time.
